internal/cmd/commands/server/server.go

```diff
@@ -341,6 +343,42 @@ func (c *Command) Run(args []string) int {
 			c.UI.Error(fmt.Errorf("Error connecting to database: %w", err).Error())
 			return 1
 		}
+
+		ctx := context.TODO()
+		sMan, err := schema.NewManager(ctx, "postgres", c.Database.DB())
+		if err != nil {
+			c.UI.Error(fmt.Errorf("Can't get schema manager: %w.", err).Error())
+			return 1
+		}
+		// This is an advisory locks on the DB which is released when the db session ends.
+		if err := sMan.SharedLock(ctx); err != nil {
+			c.UI.Error(fmt.Errorf("Unable to gain shared access to the database: %w", err).Error())
+			return 1
+		}
+		ckState, err := sMan.CurrentState(ctx)
+		if err != nil {
+			c.UI.Error(fmt.Errorf("Error checking schema state: %w", err).Error())
+			return 1
+		}
+		if !ckState.InitializationStarted {
+			c.UI.Error("Database has not been initialized. Please run `boundary database init`.")
+			return 1
+		}
+		if ckState.Dirty {
```

Comment:
Just curious, when would this happen, vs. the transaction containing the migration rolling back?

Reply:
We currently never rollback.  If there is an error we stop where it failed and leave the bit dirty.

Reply:
Wouldn't the transaction failing cause the changes from that migration to roll back? Isn't each migration its own transaction?

Reply:
That's correct, but that is only because the content of the migration issued the rollback and not because the migration library itself enforces it.   This does get close to a topic I plan on trying to fix soon.  When we fail after the the schema update has completed successfully but before all the default resources are created.  IMO we should still mark the migration as being in a dirty state.  Maybe the solution is pulling the dirty bit out to be managed through an interface in the schema manager so the caller can decide if the migration is `dirty` or not.  I'm happy to talk more about this face to face or on slack.